src/fastcs/connections/ip_connection.py

```diff
@@ -1,5 +1,41 @@
 import asyncio
+import codecs
 from dataclasses import dataclass
+from typing import Any, Callable, Coroutine, cast
+
+_AsyncFuncType = Callable[..., Coroutine[Any, Any, Any]]
+
+
+def _with_lock(func: _AsyncFuncType) -> _AsyncFuncType:
+    async def with_lock(*args: Any, **kwargs: Any) -> None:
+        self = args[0]
+        async with self.lock:
+            await func(*args, **kwargs)
+
+    return cast(_AsyncFuncType, with_lock)
```

Comment:
The old implementation wrapped all the contents of the functions in the context manager though. I can see why this would be nice for if only part of the function was in the CM, but I feel like the decorator approach is cleaner if that isn't the case.